irpc.py

```diff
@@ -35,17 +36,31 @@ def hoist_declaration(main: FuncDef,
                     main.insert(0, node)
                     break
 
+def hoist_prov_memo(main: FuncDef,
+                    provdef: ProvDef):
+
+    _name_ = f"{e}_provided"
+    memo_node = Decl(name=_name_,
+        quals=[], storage=[],funcspec=[],
+        type=TypeDecl(declname=_name_, quals=[],
+                    type=IdentifierType(names=['bool'])),
+        init=ID(name='False'),
+        bitsize=None
+        )
+    main.insert(1, memo_node)
+
 def add_provider_call(funcdef: FuncDef,
                       entnames: Set[Entity]):
 
     if is_provider(funcdef):
         entnames = entnames - set([provider_name(funcdef)])
-                
+
     # Insert the provider call
     for e, l_compound in entity2Compound(funcdef.body, entnames).items():
           for compound in l_compound:
-                f = FuncCall(name=ID(name=f'provide_{e}'), args=None)
-                compound.block_items.insert(0, f)
+                prov_if = If(cond=UnaryOp(op='!', expr=ID(name=f'{e}_provided')), iftrue=Compound(block_items=(FuncCall(name=ID(name=f'provide_{e}'), args=None), Assignment(op='=',lvalue=ID(name=f'{e}_provided'), rvalue=Constant(type='bool', value='true')))), iffalse=None)
+                compound.block_items.insert(0, prov_if)
```

Comment:
Please create a variable for `'{e}_provided`.
Better create a function/Class who take a name, and generate the name of the boolean variable.

Let say that in the future we will want to change `{e}_provided` to something else. RIght now, we need to change the value in 2 functions and 3 different places. Is not praticial